### PR TITLE
Added URL Sanitizer service and Copy Clean URL item for links

### PR DESCRIPTION
**Pull request by spylogsster**
Resolves https://github.com/brave/brave-browser/issues/23315

MVP implementation for url sanitization, spec https://docs.google.com/document/d/1ea7eF3s0WTsTlmMPr8LYp5KUyNg8eK4omjizo27qvJE/edit#

- Added a service that removes url parameters from query according to the rules shipped in resources.
- Added an item to the sharing hub menu to copy clean url
image
- Added an item to the address bar context menu to copy clean url
image

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [x] New files have MPL-2.0 license header
- [x] Adequate test coverage exists to prevent regressions
- [x] Major classes, functions and non-trivial code blocks are well-commented
- [x] Changes in component dependencies are properly reflected in `gn`
- [x] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [x] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

- Go to any url that meet config's rules and copy clean link.
- Check copied link should have removed parameters accoiring to the rules.

### CODE REVIEW

File: components/url_sanitizer/browser/url_sanitizer_service.cc [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+void URLSanitizerService::UpdateMatchers(
+    base::flat_set<std::unique_ptr<URLSanitizerService::MatchItem>> mappings) {
+  matchers_ = std::move(mappings);
+  if (initialization_callback_for_testing_)
+    std::move(initialization_callback_for_testing_).Run();
+}
+
+GURL URLSanitizerService::SanitizeURL(const GURL& initial_url) {
+  if (matchers_.empty())
+    return GURL();
```
Comment (fmarier) [inline components/url_sanitizer/browser/url_sanitizer_service.cc:114]: If we have an empty list, then I guess we don't have anything to filter out and should return `initial_url`?

File: components/url_sanitizer/browser/url_sanitizer_service.cc [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  GURL url = initial_url;
+  for (const auto& it : matchers_) {
+    if (!it->include.MatchesURL(url) || it->exclude.MatchesURL(url))
+      continue;
+    std::string debug;
```
Comment (fmarier) [inline components/url_sanitizer/browser/url_sanitizer_service.cc:119]: Looks like that variable is no longer used.

File: components/url_sanitizer/browser/url_sanitizer_service.cc [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    }
+    auto sanitized_query = StripQueryParameter(url.query(), it->params);
+    GURL::Replacements replacements;
+    replacements.SetQueryStr(sanitized_query);
+    url = url.ReplaceComponents(replacements);
+  }
+  return url;
+}
+
+std::string URLSanitizerService::StripQueryParameter(
```
Comment (fmarier) [inline components/url_sanitizer/browser/url_sanitizer_service.cc:131]: I'd suggest a comment along the lines of:
```
// FIXME: merge with browser/net/brave_site_hacks_network_delegate_helper.cc::StripQueryParameter()
```
just so we don't lose track of this future cleanup.

File: components/url_sanitizer/browser/url_sanitizer_service_unittest.cc [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+ private:
+  base::test::TaskEnvironment task_environment_;
+};
+
+TEST_F(URLSanitizerServiceUnitTest, StripQueryParameter) {
+  auto list = base::MakeFlatSet<std::string>(
+      std::vector<std::string>{"fbclid", "second"});
+
+  EXPECT_EQ(StripQueryParameter("fbclid=11&param1=1&second=2", list),
```
Comment (fmarier) [inline components/url_sanitizer/browser/url_sanitizer_service_unittest.cc:77]: I'd suggest a more complicated test to ensure that we don't break edge cases in the future:
- Before:`fbclid=11&fbclid2=ok&&param1=1&foo;bar=yes&second=2`
- After: `fbclid2=ok&&param1=1&foo;bar=yes`

File: components/url_sanitizer/browser/url_sanitizer_service_unittest.cc [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+          "utm_content=removethis&e=&f=g&=end")),
+      GURL("https://dev-pages.bravesoftware.com/clean-urls/"
+           "?brave_testing3=keep&&;b&d&e=&f=g&=end"));
+
+  WaitInitialization(R"([
+    { "include": [ "*://*/*"], "params": ["query"] }
+  ])");
+
+  EXPECT_EQ(SanitizeURL(GURL("https://brave.com/?query=removethis")),
+            GURL("https://brave.com/?"));
```
Comment (fmarier) [inline components/url_sanitizer/browser/url_sanitizer_service_unittest.cc:115]: In this case we should ideally delete the query string and end up with just `https://brave.com/` (no question mark at the end).

File: components/url_sanitizer/browser/url_sanitizer_service_unittest.cc [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+  EXPECT_EQ(SanitizeURL(GURL("http://twitter.com/post/?query=removethis")),
+            GURL("http://twitter.com/post/?"));
+
+  EXPECT_EQ(SanitizeURL(GURL("chrome://settings/?query=donotremovethis")),
+            GURL("chrome://settings/?query=donotremovethis"));
+
+  WaitInitialization(R"([
```
Comment (fmarier) [inline components/url_sanitizer/browser/url_sanitizer_service_unittest.cc:123]: These tests are great!

**Review by fmarier — CHANGES_REQUESTED** [review]
Which contextual menu is this hooking into? It looks like it might be the one you get when you right-click on a link. If that's the case, then that was the original idea (the screenshot in the issue), but I ended up changing this in the spec because I realized that these URLs are not fully clean already and so they're more complicated to handle.

What I suggest instead in the spec is to hook into the contextual menu you get when you right-click on the **address in the URL bar**. That URL will have gone through the debouncer and the regular query filter and so all that's left to remove is the contents of the new JSON list. We can leave the [link contextual menu for a later phase](https://docs.google.com/document/d/1ea7eF3s0WTsTlmMPr8LYp5KUyNg8eK4omjizo27qvJE/edit#heading=h.utvf3nnnswm0).

File: components/url_sanitizer/browser/url_sanitizer_service_unittest.cc (hunk 4 above)
Comment (spylogsster) [reply components/url_sanitizer/browser/url_sanitizer_service_unittest.cc:77]: added it too

File: components/url_sanitizer/browser/url_sanitizer_service.cc (hunk 3 above)
Comment (spylogsster) [reply components/url_sanitizer/browser/url_sanitizer_service.cc:132]: done

**Comment by spylogsster** [comment]
> Which contextual menu is this hooking into? It looks like it might be the one you get when you right-click on a link. If that's the case, then that was the original idea (the screenshot in the issue), but I ended up changing this in the spec because I realized that these URLs are not fully clean already and so they're more complicated to handle.
> 
> What I suggest instead in the spec is to hook into the contextual menu you get when you right-click on the **address in the URL bar**. That URL will have gone through the debouncer and the regular query filter and so all that's left to remove is the contents of the new JSON list. We can leave the [link contextual menu for a later phase](https://docs.google.com/document/d/1ea7eF3s0WTsTlmMPr8LYp5KUyNg8eK4omjizo27qvJE/edit#heading=h.utvf3nnnswm0).

done

File: components/url_sanitizer/browser/url_sanitizer_service_unittest.cc (hunk 5 above)
Comment (spylogsster) [reply components/url_sanitizer/browser/url_sanitizer_service_unittest.cc]: done

File: components/url_sanitizer/browser/url_sanitizer_service.cc (hunk 2 above)
Comment (spylogsster) [reply components/url_sanitizer/browser/url_sanitizer_service.cc]: removed, thanks

File: components/url_sanitizer/browser/url_sanitizer_service.cc (hunk 1 above)
Comment (spylogsster) [reply components/url_sanitizer/browser/url_sanitizer_service.cc]: done

File: app/brave_generated_resources.grd [hunk 7]
```diff
@@ -922,7 +922,9 @@ Or change later at <ph name="SETTINGS_EXTENIONS_LINK">$2<ex>brave://settings/ext
       <message name="IDS_IMPORTED_FROM_BOOKMARK_FOLDER" desc="Name for bookmark panel folder imported from another browser">
         Imported from <ph name="BROWSER_NAME">$1<ex>Chrome</ex></ph>.
       </message>
-
+      <message name="IDS_COPY_CLEAN_LINK" desc="The name of the context menu item to copy clean link">
+        Copy clean link address
```
Comment (fmarier) [inline app/brave_generated_resources.grd:926]: We can probably just use "Copy clean link" (no `address`) since the original option is called "Copy link".

File: components/url_sanitizer/resources/default_trackers.json [hunk 8]
```diff
@@ -0,0 +1,64 @@
+[
```
Comment (fmarier) [inline components/url_sanitizer/resources/default_trackers.json:1]: I guess we should be able to delete this file now?

File: app/brave_generated_resources.grd (hunk 7 above)
Comment (fmarier) [reply app/brave_generated_resources.grd:926]: After discussing with @rmcfadden3 , let's change this to: "Copy clean link (only main URL)"

**Comment by rmcfadden3** [comment]
I like it!

File: components/url_sanitizer/resources/default_trackers.json (hunk 8 above)
Comment (spylogsster) [reply components/url_sanitizer/resources/default_trackers.json]: done

File: app/brave_generated_resources.grd (hunk 7 above)
Comment (spylogsster) [reply app/brave_generated_resources.grd]: done

**Review by fmarier — APPROVED** [review]

File: browser/brave_browser_process.h [hunk 9]
```diff
@@ -87,6 +88,8 @@ class BraveBrowserProcess {
 #endif
   virtual debounce::DebounceComponentInstaller*
   debounce_component_installer() = 0;
+  virtual brave::URLSanitizerComponentInstaller*
+  url_sanitizer_component_installer() = 0;
```
Comment (cdesouza-chromium) [inline browser/brave_browser_process.h:92]: Nit: Fixing this might not be possible other functions around are violating the style too, but these virtual functions should be using `PascalCase` if we were to follow chromium style here, since `unix_case` is used in chromium to indicate that a function is inlined when called, which won't be the case for these virtuals.

File: components/url_sanitizer/browser/url_sanitizer_component_installer.h [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+      brave_component_updater::LocalDataFilesService* local_data_files_service);
+  URLSanitizerComponentInstaller(const URLSanitizerComponentInstaller&) =
+      delete;
+  URLSanitizerComponentInstaller& operator=(
+      const URLSanitizerComponentInstaller&) = delete;
+  ~URLSanitizerComponentInstaller() override;
+
+  class Observer : public base::CheckedObserver {
+   public:
+    virtual void OnRulesReady(const std::string&) = 0;
```
Comment (cdesouza-chromium) [inline components/url_sanitizer/browser/url_sanitizer_component_installer.h:41]: Nit: Having the name of the argument could help clarity of the observer here.

File: chromium_src/chrome/browser/sharing_hub/sharing_hub_model.cc [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "chrome/browser/sharing_hub/sharing_hub_model.h"
+#include "brave/app/brave_command_ids.h"
+#include "brave/grit/brave_generated_resources.h"
+#include "chrome/grit/generated_resources.h"
+
+int GetOriginalSharingHubCopyLinkID() {
```
Comment (goodov) [inline chromium_src/chrome/browser/sharing_hub/sharing_hub_model.cc:11]: should be static or in anonymous ns.
also can be replaced with `static auto IDS_SHARING_HUB_COPY_LINK_LABEL_ORIGINAL = IDS_SHARING_HUB_COPY_LINK_LABEL;`

File: components/url_sanitizer/browser/url_sanitizer_component_installer.h [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+  };
+
+  // Implementation of brave_component_updater::LocalDataFilesObserver
+  void OnComponentReady(const std::string& component_id,
+                        const base::FilePath& install_dir,
+                        const std::string& manifest) override;
+
+  void AddObserver(Observer* observer) { observers_.AddObserver(observer); }
```
Comment (cdesouza-chromium) [inline components/url_sanitizer/browser/url_sanitizer_component_installer.h:49]: Is there any reason in particular why these functions have their implementation in the header file? Unless there's some reason in particular, I think it would be better to have their implementation moved out of the header.

File: components/url_sanitizer/browser/url_sanitizer_service.cc [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+  }
+  return valid;
+}
+
+bool CreateParamsList(const base::Value* value,
+                      base::flat_set<std::string>* params) {
+  if (!value->is_list() || !params)
+    return false;
+
+  for (const auto& it : value->GetList()) {
```
Comment (cdesouza-chromium) [inline components/url_sanitizer/browser/url_sanitizer_service.cc:43]: `DCHECK(it.is_string());`?

Nit: `const auto& param`?

File: components/url_sanitizer/browser/url_sanitizer_service.cc [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+    params->insert(it.GetString());
+  }
+  return true;
+}
+
+base::flat_set<std::unique_ptr<URLSanitizerService::MatchItem>> ParseFromJson(
+    const std::string& json) {
+  auto parsed_json = base::JSONReader::ReadAndReturnValueWithError(json);
+  if (!parsed_json.has_value() || !parsed_json.value().is_list()) {
```
Comment (cdesouza-chromium) [inline components/url_sanitizer/browser/url_sanitizer_service.cc:52]: Nit: I think the following would be slightly easier to read

```
if (!parsed_json.has_value() || !parsed_json->is_list())
```

File: components/url_sanitizer/browser/url_sanitizer_service.cc [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+    VLOG(1) << "Error parsing feature JSON: " << parsed_json.error().message;
+    return {};
+  }
+  base::flat_set<std::unique_ptr<URLSanitizerService::MatchItem>> matchers;
+  for (const auto& it : parsed_json.value().GetList()) {
```
Comment (cdesouza-chromium) [inline components/url_sanitizer/browser/url_sanitizer_service.cc:57]: `->`

File: components/url_sanitizer/browser/url_sanitizer_service.cc [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+    auto* include_list = it.FindListPath("include");
```
Comment (cdesouza-chromium) [inline components/url_sanitizer/browser/url_sanitizer_service.cc:58]: Please create an instance of `base::Value::Dict` and use `Find` (do not use `FindList`, as you want a `base::Value` here) to look up this value, as `FindListPath` is deprecated.

File: components/url_sanitizer/browser/url_sanitizer_service.cc [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+    if (!include_list)
+      continue;
+    extensions::URLPatternSet include_matcher;
+    if (!CreateURLPatternSetFromValue(include_list, &include_matcher))
+      continue;
+    auto* params_list = it.FindListPath("params");
```
Comment (cdesouza-chromium) [inline components/url_sanitizer/browser/url_sanitizer_service.cc:64]: `base::FindListPath` -> `base::Value::Dict::Find`

File: components/url_sanitizer/browser/url_sanitizer_service.cc [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+  bool valid = result->Populate(
+      base::Value::AsListValue(*value),
+      URLPattern::SCHEME_HTTP | URLPattern::SCHEME_HTTPS, false, &error);
+  if (!valid) {
+    VLOG(1) << "Unable to create url pattern:" << error;
+  }
+  return valid;
+}
+
+bool CreateParamsList(const base::Value* value,
```
Comment (cdesouza-chromium) [inline components/url_sanitizer/browser/url_sanitizer_service.cc:38]: I think we could have `absl::optional<base::flat_set<std::string>>` as the return type, rather than using an out param, which is discouraged.

File: components/url_sanitizer/browser/url_sanitizer_service.cc [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+      continue;
+    auto* params_list = it.FindListPath("params");
+    base::flat_set<std::string> params;
+    if (!CreateParamsList(params_list, &params))
+      continue;
+
+    extensions::URLPatternSet exclude_matcher;
+    CreateURLPatternSetFromValue(it.FindListPath("exclude"), &exclude_matcher);
+    auto item = std::make_unique<URLSanitizerService::MatchItem>(
+        std::move(include_matcher), std::move(exclude_matcher), params);
```
Comment (cdesouza-chromium) [inline components/url_sanitizer/browser/url_sanitizer_service.cc:72]: `std::move(param)`

File: components/url_sanitizer/browser/url_sanitizer_service.cc [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+URLSanitizerService::URLSanitizerService() = default;
+
+URLSanitizerService::~URLSanitizerService() = default;
+
+URLSanitizerService::MatchItem::MatchItem() = default;
+URLSanitizerService::MatchItem::~MatchItem() = default;
+
+URLSanitizerService::MatchItem::MatchItem(extensions::URLPatternSet in,
+                                          extensions::URLPatternSet ex,
+                                          base::flat_set<std::string> prm) {
```
Comment (cdesouza-chromium) [inline components/url_sanitizer/browser/url_sanitizer_service.cc:91]: Is it possible to have these members initialised in the initialiser list?

```
  : include(std::move(in)), exclude(std::move(ex)), params(std::move(prm))
```

File: components/url_sanitizer/browser/url_sanitizer_service.cc [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+    const std::string& query,
+    const base::flat_set<std::string>& trackers) {
+  const std::vector<std::string> input_kv_strings =
+      SplitString(query, "&", base::KEEP_WHITESPACE, base::SPLIT_WANT_ALL);
+  std::vector<std::string> output_kv_strings;
+  int disallowed_count = 0;
+  for (const std::string& kv_string : input_kv_strings) {
+    const std::vector<std::string> pieces = SplitString(
+        kv_string, "=", base::KEEP_WHITESPACE, base::SPLIT_WANT_NONEMPTY);
+    const std::string& key = pieces.empty() ? "" : pieces[0];
```
Comment (cdesouza-chromium) [inline components/url_sanitizer/browser/url_sanitizer_service.cc:146]: Prefer using `std::string()` rather than `""`, as that results in better codegen.

File: components/url_sanitizer/browser/url_sanitizer_service.h [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+  void Initialize(const std::string& json);
+  void OnRulesReady(const std::string&) override;
+
+ protected:
+  friend class URLSanitizerServiceUnitTest;
+
+  void UpdateMatchers(
+      base::flat_set<std::unique_ptr<URLSanitizerService::MatchItem>>);
+
+  std::string StripQueryParameter(const std::string& query,
```
Comment (cdesouza-chromium) [inline components/url_sanitizer/browser/url_sanitizer_service.h:58]: Do you think it would be possible to add some comment on top explaining what this function does?

Comment (spylogsster) [reply components/url_sanitizer/browser/url_sanitizer_service.h:58]: done

File: components/url_sanitizer/browser/url_sanitizer_service.cc (hunk 21 above)
Comment (spylogsster) [reply components/url_sanitizer/browser/url_sanitizer_service.cc]: done

File: components/url_sanitizer/browser/url_sanitizer_service.cc (hunk 20 above)
Comment (spylogsster) [reply components/url_sanitizer/browser/url_sanitizer_service.cc]: done

File: components/url_sanitizer/browser/url_sanitizer_service.cc (hunk 19 above)
Comment (spylogsster) [reply components/url_sanitizer/browser/url_sanitizer_service.cc]: done

File: components/url_sanitizer/browser/url_sanitizer_service.cc (hunk 18 above)
Comment (spylogsster) [reply components/url_sanitizer/browser/url_sanitizer_service.cc]: done

File: components/url_sanitizer/browser/url_sanitizer_service.cc (hunk 17 above)
Comment (spylogsster) [reply components/url_sanitizer/browser/url_sanitizer_service.cc]: done

File: components/url_sanitizer/browser/url_sanitizer_service.cc (hunk 16 above)
Comment (spylogsster) [reply components/url_sanitizer/browser/url_sanitizer_service.cc]: done

File: components/url_sanitizer/browser/url_sanitizer_service.cc (hunk 15 above)
Comment (spylogsster) [reply components/url_sanitizer/browser/url_sanitizer_service.cc]: done

File: components/url_sanitizer/browser/url_sanitizer_service.cc (hunk 14 above)
Comment (spylogsster) [reply components/url_sanitizer/browser/url_sanitizer_service.cc]: done

File: chromium_src/chrome/browser/ui/views/omnibox/omnibox_view_views.cc [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+#include "chrome/browser/ui/views/location_bar/location_bar_view.h"
+#include "chrome/browser/ui/views/omnibox/omnibox_view_views.h"
+#include "chrome/browser/ui/views/send_tab_to_self/send_tab_to_self_bubble_controller.h"
+#include "chrome/grit/generated_resources.h"
+#include "components/omnibox/browser/omnibox_edit_model.h"
+#include "components/strings/grit/components_strings.h"
+#include "ui/base/clipboard/scoped_clipboard_writer.h"
+#include "ui/views/controls/textfield/textfield.h"
+
+void BraveUpdateContextMenu(ui::SimpleMenuModel* menu_contents) {
```
[the remark this answers is not in the document]
Comment (spylogsster) [reply chromium_src/chrome/browser/ui/views/omnibox/omnibox_view_views.cc:27]: done

File: chromium_src/chrome/browser/sharing_hub/sharing_hub_model.cc (hunk 11 above)
Comment (spylogsster) [reply chromium_src/chrome/browser/sharing_hub/sharing_hub_model.cc:13]: done, thanks

File: components/url_sanitizer/browser/url_sanitizer_component_installer.h (hunk 12 above)
Comment (spylogsster) [reply components/url_sanitizer/browser/url_sanitizer_component_installer.h]: done

File: components/url_sanitizer/browser/url_sanitizer_component_installer.h (hunk 10 above)
Comment (spylogsster) [reply components/url_sanitizer/browser/url_sanitizer_component_installer.h]: done

File: browser/brave_browser_process.h (hunk 9 above)
Comment (spylogsster) [reply browser/brave_browser_process.h]: done, thanks

**Review by cdesouza-chromium — APPROVED** [review]
LGTM+++

File: browser/url_sanitizer/url_sanitizer_service_factory.cc [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+    g_brave_browser_process->URLSanitizerComponentInstaller()->AddObserver(
+        service);
+  }
+  return service;
+}
+
+bool URLSanitizerServiceFactory::ServiceIsNULLWhileTesting() const {
+  return false;
+}
+
```
Comment (goodov) [inline browser/url_sanitizer/url_sanitizer_service_factory.cc:53]: seems like this service is fine to use as is in incognito profiles. no need to create another instance.
```
content::BrowserContext* URLSanitizerServiceFactory::GetBrowserContextToUse(
      content::BrowserContext* context) const {
  return chrome::GetBrowserContextRedirectedInIncognito(context);
}
```

Comment (spylogsster) [reply browser/url_sanitizer/url_sanitizer_service_factory.cc:53]: done, thanks!

File: components/url_sanitizer/browser/url_sanitizer_component_installer.cc [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+#include "brave/components/brave_component_updater/browser/dat_file_util.h"
+#include "brave/components/brave_component_updater/browser/local_data_files_service.h"
+#include "net/base/registry_controlled_domains/registry_controlled_domain.h"
+
+using brave_component_updater::LocalDataFilesObserver;
+using brave_component_updater::LocalDataFilesService;
+
+namespace brave {
+
+const char kCleanURLsConfigFile[] = "clean-urls.json";
```
Comment (bsclifton) [inline components/url_sanitizer/browser/url_sanitizer_component_installer.cc:28]: So this installer is piggybacking off the `Brave Local Data Updater` component? (used for debounce list, etc)

Where does the `clean-urls.json` live? (in which repo?). Curious if an update was made, what steps would have to be taken to push new package out

Comment (spylogsster) [reply components/url_sanitizer/browser/url_sanitizer_component_installer.cc:28]: https://github.com/brave/brave-core-crx-packager/pull/464, it downloaded by `Brave Local Data Updater` component

Comment (fmarier) [reply components/url_sanitizer/browser/url_sanitizer_component_installer.cc:28]: https://github.com/brave/adblock-lists/blob/master/brave-lists/clean-urls.json is where the list lives

Comment (fmarier) [reply components/url_sanitizer/browser/url_sanitizer_component_installer.cc:28]: Updates are supposed to be automatic, but there's currently a problem with that (https://github.com/brave/adblock-lists/issues/955) and I need to look into it with @antonok-edm.

Comment (bsclifton) [reply components/url_sanitizer/browser/url_sanitizer_component_installer.cc:28]: This is great- thanks! 😄

Comment (bsclifton) [reply components/url_sanitizer/browser/url_sanitizer_component_installer.cc:28]: @fmarier does adding this new `clean-urls.json` close out https://github.com/brave/brave-browser/issues/10188?

Comment (bsclifton) [reply components/url_sanitizer/browser/url_sanitizer_component_installer.cc:28]: I guess this can be done in another PR, but it would be cool to remove the existing hardcoded rules (ex: https://github.com/brave/brave-core/blob/d2d6b59666afa92b26e8c207ee936aad0a1c19f1/browser/net/brave_site_hacks_network_delegate_helper.cc#L31-L72) and then use this new keyed service also (`URLSanitizerService::SanitizeURL`)

**Comment by bsclifton** [comment]
I have the code checked out locally and built - but it doesn't seem to be working

Here are the steps I tried:
1. Build code / launch
2. Visit https://dev-pages.bravesoftware.com/clean-urls/?brave_testing1=foo&brave_testing2=bar&brave_testing3=keep&&;b&d&utm_content=removethis&e=&f=g&=end
3. Site doesn't resolve, but that's OK
4. Right click omnibox and pick ``
    image
5. Paste the clipboard into Notepad or similar
6. It still has the `brave_testing1=foo` and `brave_testing2=bar` (and other params)

**Comment by bsclifton** [comment]
OK seems `clean-urls` only has the contents `[]`. 

I opened the file manually (`%USERPROFILE%\AppData\local\BraveSoftware\Brave-Browser-Development\User Data\afalakplffnnnlkncjhbmahjfjhmlkal\1.0.188\1\clean-urls.json`) and paste in the contents from https://github.com/brave/adblock-lists/blob/master/brave-lists/clean-urls.json

This worked great as expected 👍 
https://github.com/brave/adblock-lists/blob/master/brave-lists/clean-urls.json

**Review by bsclifton — APPROVED** [review]
Left comments - maybe we can use https://github.com/brave/brave-browser/issues/10188 as the follow up issue?

The list (matching the spec) doesn't seem to be live yet. But as shared, I tested by manually pasting the list from https://github.com/brave/adblock-lists/blob/master/brave-lists/clean-urls.json and it works great

File: components/url_sanitizer/browser/url_sanitizer_component_installer.cc (hunk 25 above)
Comment (fmarier) [reply components/url_sanitizer/browser/url_sanitizer_component_installer.cc:28]: > @fmarier does adding this new `clean-urls.json` close out [brave/brave-browser#10188](https://github.com/brave/brave-browser/issues/10188)?

No, that would be in Phase 5 of https://docs.google.com/document/d/1ea7eF3s0WTsTlmMPr8LYp5KUyNg8eK4omjizo27qvJE/edit#heading=h.zfmdqhvgwdme

Comment (fmarier) [reply components/url_sanitizer/browser/url_sanitizer_component_installer.cc:28]: @clifton Merging these two lists is something we've considered, but there are a few trade-offs and so we settled on an iterative approach (as written in the spec above). This PR is just Phase 1.

**Comment by fmarier** [comment]
> Left comments - maybe we can use [brave/brave-browser#10188](https://github.com/brave/brave-browser/issues/10188) as the follow up issue?

I think we should file a separate follow-up issue because we could merge the parsing code between these two features without waiting for the lists to get merged, which is going to be trickier.

**Comment by goodov** [comment]
you need to do `WIPE_WORKSPACE`. just a rebuild won't work, because there're some dirs left that are not removed automatically.

**Comment by bsclifton** [comment]
Created https://github.com/brave/brave-browser/issues/25104 as the follow up for the refactoring 😄👍

**Review by goodov — APPROVED** [review]
